---
title: 'PASS1B BIC QC report: RNA-seq'
author: "David Amar, Nicole Gay"
date: "05/03/2020"
output:
  html_document:
    df_print: paged
geometry: margin=2cm
---

```{r setup, include=FALSE}
library(knitr)
library(kableExtra)
knitr::opts_chunk$set(echo = TRUE)
# knitr::opts_chunk$set(dev = 'jpeg') # helps in reducing the pdf size
```

In this document we create different QC reports, including: (1) correlations between principal components and other technical or biological variables, denoted as the metadata variables, (2) correlations among the metadata variables, and (3) outlier detection. These tests are done separately for each dataset - a set of samples from a single site and a single tissue.

The RNA-seq analysis has two additional tests that are not a part of the standatd tests that we perform automatically for each assay: (1) sex checks, and (2) correlations between freeze times and sex.

# Load the data and set the session

```{r,message=FALSE,warning=FALSE}
library(data.table);library(ggplot2);library(ggcorrplot);library(caret)
# load the data and functions to the session
repo_dir = '/oak/stanford/groups/smontgom/nicolerg/src/MOTRPAC/'
data_dir = '/oak/stanford/groups/smontgom/nicolerg/MOTRPAC/PASS_ANALYSIS/PASS1B/RNA/differential_splicing/outputs/perind_numers_counts'
source(sprintf("%s/motrpac-bic-norm-qc/tools/unsupervised_normalization_functions.R", repo_dir))
source(sprintf("%s/motrpac-bic-norm-qc/tools/gcp_functions.R", repo_dir))
source(sprintf("%s/motrpac-bic-norm-qc/tools/qc_report_aux_functions.R", repo_dir))
source(sprintf("%s/motrpac-bic-norm-qc/tools/association_analysis_methods.R", repo_dir))
source(sprintf("%s/motrpac-mawg/pass1b-06/integrative/outliers_covariates/pi1_cook_fx.R", repo_dir))
source(sprintf("%s/motrpac-mawg/pass1b-06/figures/motrpac_colors_abbr.R", repo_dir))

# # Data buckets
# main_bucket1 = "gs://motrpac-portal-transfer-stanford/Output/PASS1B/RNA-SEQ/batch3_20191016/results/"
# main_bucket2 = "gs://motrpac-portal-transfer-sinai/Output/rna-seq/rat/"
# buckets = data.frame(
#   site = c("stanford","sinai","sinai"),
#   bucket = c(main_bucket1,paste0(main_bucket2,"batch5_20191031/results/"),
#              paste0(main_bucket2,"batch4_20191021/results/")),
#   qc_file = c(
#     "rnaseq_pipeline_qc_metrics_batch3.csv",
#     "rnaseq_pipeline_qc_metrics_batch5.csv",
#     "rnaseq_pipeline_qc_metrics_batch4.csv"
#   ),
#   counts_file = rep("rsem_genes_count.txt",3),
#   fpkms_file = rep("rsem_genes_fpkm.txt",3),
#   stringsAsFactors = F
# )

# loading splicing results from local files for now
rnaseq_pipeline_data = list()
for(j in system(sprintf('ls %s',data_dir), intern=T)){
  file = sprintf('%s/%s', data_dir, j)
  tissue = gsub("_perind.*","",j)
  curr_counts = read.table(file, sep=' ', header=T, check.names=F)
  rnaseq_pipeline_data[[tissue]] = list(counts = curr_counts)
}

# RNA-seq QC 
qc = dl_read_gcp("gs://motrpac-internal-release2-results/pass1b-06/transcriptomics/qa-qc/motrpac_pass1b-06_transcript-rna-seq_qa-qc-metrics.csv", sep=',')

# Metadata from DMAQC: use the merged data frame
dmaqc_metadata = 'gs://motrpac-internal-release2-results/pass1b-06/phenotype/motrpac_pass1b-06_pheno_viallabel-data.txt'
dmaqc_dict = 'gs://motrpac-internal-release2-results/pass1b-06/phenotype/motrpac_pass1b_pheno_merged-dictionary.txt'

# download and format phenotypic data 
dmaqc_metadata = dl_read_gcp(dmaqc_metadata, sep='\t')
cols = dl_read_gcp(dmaqc_dict, sep='\t')
old_cols = colnames(dmaqc_metadata)
new_cols = tolower(cols[match(old_cols, BICUniqueID), FullName]) 
colnames(dmaqc_metadata) = new_cols # this isn't perfect, but we don't care about the columns it doesn't work for for now

# make some variables human-readable
# create new variables "protocol", "agegroup", "intervention", "sacrificetime", "sex" with readable strings 
for (var in c('key.protocol','key.agegroup','key.intervention','key.sacrificetime','registration.sex')){
  d = cols[Field.Name == gsub('.*\\.','',var)]
  keys=unname(unlist(strsplit(d[,Categorical.Values],'\\|')))
  values=tolower(unname(unlist(strsplit(d[,Categorical.Definitions],'\\|'))))
  names(values) = keys
  # match keys to values; create new column 
  new_var = gsub(".*\\.","",var)
  dmaqc_metadata[,(new_var) := unname(values)[match(get(var), names(values))]]
}
dmaqc_metadata[,time_to_freeze := calculated.variables.frozetime_after_train - calculated.variables.deathtime_after_train]
# clean up "sacrificetime"
dmaqc_metadata[,sacrificetime := sapply(sacrificetime, function(x) gsub(' week.*','W',x))]
# clean up 'intervention'
dmaqc_metadata[grepl('training',intervention), intervention := 'training']
# make "group" - "1w", "2w", "4w", "8w", "control"
dmaqc_metadata[,group := sacrificetime]
dmaqc_metadata[intervention == 'control', group := 'SED']
# make tech ID a string
dmaqc_metadata[,specimen.processing.techid := paste0('tech',specimen.processing.techid)]
dmaqc_metadata[,tissue := specimen.processing.sampletypedescription]

# merge qc
rnaseq_meta = merge(dmaqc_metadata, qc, by.x='viallabel', by.y='vial_label')
rnaseq_meta = as.data.frame(rnaseq_meta)
colnames(rnaseq_meta) = gsub(" .*","",colnames(rnaseq_meta))
rownames(rnaseq_meta) = rnaseq_meta$viallabel
rnaseq_meta$animal.key.is_control = rnaseq_meta$key.intervention == 3
```

# Metadata variables

MOP-flagging samples is skipped here because it's redundant with the pass1b_rnaseq QC report. 

We next define the set of metadata variables. The first set is the set of variables from the data generating pipeline. The second set is of the biospecimens and is shared across all assays.

```{r}
# Define technical and biological variables to be considered in the QC
pipeline_qc_cols = c("RIN","reads",
                     "Lib_frag_size",
                     "RNA_extr_conc",
                     "pct_rRNA","pct_globin",
                     "pct_umi_dup","median_5_3_bias",
                     "pct_multimapped","pct_GC","pct_chrM")
biospec_cols = c("registration.sex","key.sacrificetime",
                  "animal.key.is_control",
                  "terminal.weight.bw","time_to_freeze",
                  "bid","pid","group","viallabel")
```

# Dataset normalization

We use log2 normalization for the intron cluster counts. 

```{r,eval=TRUE,warning=F,message=F}
# Labels for separating the data by site and tissue
rnaseq_meta_batchs = unique(rnaseq_meta[,c("GET_site","Tissue")])
norm_rnaseq_data = list()
for(i in 1:nrow(rnaseq_meta_batchs)){
  curr_site = rnaseq_meta_batchs[i,1]
  curr_tissue = rnaseq_meta_batchs[i,2]
  curr_samples = as.character(
    rnaseq_meta$viallabel[rnaseq_meta$GET_site==curr_site &
    rnaseq_meta$Tissue==curr_tissue])
  curr_site = tolower(curr_site)
  curr_tissue = tolower(curr_tissue)
  curr_tissue = gsub("hypothalmus","hypothalamus",curr_tissue)
  # For rat data, take samples whose label id starts with "9" - remove qc pools
  curr_samples = curr_samples[grepl("^9",curr_samples)]
  
  # For normalized counts use:
  my_tissue = gsub(" ","_",gsub(" powder","",curr_tissue))
  curr_counts = rnaseq_pipeline_data[[my_tissue]]$counts
  if(is.null(curr_counts)){next}
  curr_data = log(curr_counts+1)
  
  # remove zero variance rows
  curr_data = curr_data[apply(curr_data,1,sd)>0,]
  
  # look at distn of CV (hist of stdev norm by mean)
  # if there's a big peak or tail, PCA is affected
  
  
  # at this point, the highest variance intron clusters are on sex chromosomes. should I remove sex chroms?
  curr_meta1 = rnaseq_meta[curr_samples,pipeline_qc_cols]
  curr_meta2 = rnaseq_meta[curr_samples,biospec_cols]
  non_control_inds = (!curr_meta2$key.intervention == 3)
  curr_name = paste(curr_tissue,curr_site,sep=",")
  norm_rnaseq_data[[curr_name]] = list(log2 = curr_data,
      pipeline_meta = curr_meta1,dmaqc_meta = curr_meta2,
      non_control_inds=non_control_inds)
}
save(norm_rnaseq_data, file="/oak/stanford/groups/smontgom/nicolerg/MOTRPAC/PASS_ANALYSIS/PASS1B/RNA/differential_splicing/bic_qc_report/rdata/norm_rnaseq_data.RData")
# save_to_bucket(norm_rnaseq_data,file = "norm_rnaseq_data.RData",
#               bucket = "gs://bic_data_analysis/pass1a/rnaseq/")
```

## PCA outliers

In this analysis, outliers are flagged by examining the boxplot of each PC, extending its whiskers to three times the inter-quantile range away from the boxplot. Samples outside this range are then flagged. 

```{r,out.height='50%',out.width='50%',message=FALSE,warning=FALSE}
pca_outliers_report = list()
pca_outliers = c()
for(currname in names(norm_rnaseq_data)[1:3]){
  
  print(currname)
  
  # run PCA 
  # with all samples
  pca = id_pca_outliers(norm_rnaseq_data[[currname]][["log2"]], 0.05, plot=F, verbose=F, center_scale=T, iqr_coef=3)
  curr_pca = pca$prcomp_obj
  curr_pcax = curr_pca$x[,1:5]
  explained_var = summary(curr_pca)[["importance"]][3,pca$num_pcs]
  norm_rnaseq_data[[currname]][["pca"]] = list(
    pcax = curr_pcax, 
    explained_var=explained_var
  )

  # split by sex for outlier calling
  for(s in unique(norm_rnaseq_data[[currname]][["dmaqc_meta"]]$registration.sex)){
    # subset to this sex
    dat = norm_rnaseq_data[[currname]][["log2"]]
    meta = norm_rnaseq_data[[currname]][["dmaqc_meta"]]
    meta = meta[meta$registration.sex==s,]
    vl = as.character(meta$viallabel)
    
    pca = id_pca_outliers(norm_rnaseq_data[[currname]][["log2"]][,vl], 0.05, plot=T, verbose=T, center_scale=T, iqr_coef=3)
    # outlier report
    if(!is.null(pca$pc_outliers_report)){
      pca_out = data.table(pca$pc_outliers_report)
      pca_out[,dataset := currname]
      pca_out[,sex := s]
      pca_outliers_report[[currname]] = pca_out
      pca_outliers = c(pca_outliers, pca$pca_outliers)
    }
  }
  
  # plot
  df = data.frame(
    PC1 = curr_pcax[,1],PC2 = curr_pcax[,2],
    group = norm_rnaseq_data[[currname]][["dmaqc_meta"]][,"group"],
    viallabel = norm_rnaseq_data[[currname]][["dmaqc_meta"]][,"viallabel"],
    sex = as.character(norm_rnaseq_data[[currname]][["dmaqc_meta"]][,"registration.sex"]),
    stringsAsFactors = F
  )
  df$sex[df$sex=="1"] = "F"
  df$sex[df$sex=="2"] = "M"
  df$group = factor(df$group, levels = c("SED", 
                                         "1W",
                                         "2W",
                                         "4W",
                                         "8W"))
  df$outlier = ifelse(as.character(df$viallabel) %in% as.character(pca_outliers), 1, 0)
  
  p = ggplot(df, aes(x=PC1, y=PC2, fill=group, shape=sex, colour=factor(outlier))) + 
    geom_point(size=3) +
    ggtitle(currname) + 
    theme_classic() +
    scale_fill_manual(values=group_cols) +
    scale_colour_manual(values=c('0'='black','1'='red')) +
    scale_shape_manual(values=c('M'=21, 'F'=24)) +
    guides(fill=guide_legend(override.aes = list(shape=21)),
           colour=guide_legend(override.aes = list(shape=21))) +
    labs(colour='outlier?')
    
  plot(p)
}
if(!is.null(dim(pca_outliers_report))){
  pca_outliers_report = rbindlist(pca_outliers_report)
  colnames(pca_outliers_report) =  c("PC","sample","score","dataset","sex")
}
```

# PC-based association analysis

Here we first take each tissue and analyze its samples. For the top principal components (5) we compute their association with the metadata variables. We use Spearman correlation and a linear test for significance. 

```{r,out.height='50%',out.width='50%',message=FALSE,warning=FALSE}
pcs_vs_qc_var_report = c()
metadata_variable_assoc_report = c()
p_thr = 0.0001
for(currname in names(norm_rnaseq_data)){
  curr_meta = cbind(norm_rnaseq_data[[currname]]$pipeline_meta,
                     norm_rnaseq_data[[currname]]$dmaqc_meta)
  # remove the text group description
  curr_meta = curr_meta[!grepl("randgr",names(curr_meta))]
  # remove metadata variables with NAs
  curr_meta = curr_meta[,!apply(is.na(curr_meta),2,any)]
  # remove bid and pid
  curr_meta = curr_meta[,!grepl("pid|bid|viallabel|group",names(curr_meta))]
  # take the PCA results
  curr_pcax = norm_rnaseq_data[[currname]][["pca"]][["pcax"]]
  
  corrs = cor(curr_pcax,curr_meta,method="spearman")
  corrsp = pairwise_eval(
    curr_pcax,curr_meta,func=pairwise_association_wrapper,
    f=1)
  
  # Some ggplots have to be printed to be shown in the notebook
  print(ggcorrplot(t(corrs),lab=T,lab_size=2.5,hc.order = F) +
    ggtitle(currname) +
    theme(plot.title = element_text(hjust = 0.5,size=20)))
  
  for(i in 1:nrow(corrsp)){
    for(j in 1:ncol(corrsp)){
      if(corrsp[i,j]>p_thr){next}
      pcs_vs_qc_var_report = rbind(pcs_vs_qc_var_report,
            c(currname,
              rownames(corrsp)[i],colnames(corrsp)[j],corrs[i,j],corrsp[i,j])
            )
    }
  }
  
  # compute correlations among the metadata variables
  corrs = cor(curr_meta,method="spearman")
  corrsp = pairwise_eval(
    curr_meta,func=pairwise_association_wrapper,
    f=1)
  
  # Some ggplots have to be printed to be shown in the notebook
  print(ggcorrplot(corrs,lab=T,lab_size=1.5,hc.order = T))
  
  for(n1 in rownames(corrsp)){
    for(n2 in rownames(corrsp)){
      if(n1==n2){break}
      if(n1 %in% biospec_cols &&
         n2 %in% biospec_cols) {next}
      if(corrsp[n1,n2]>p_thr){next}
      metadata_variable_assoc_report = rbind(metadata_variable_assoc_report,
            c(currname,n1,n2,corrs[n1,n2],corrsp[n1,n2])
            )
    }
  }
}
# Format the reports - for a nicer presentation in a table
colnames(metadata_variable_assoc_report) = c(
  "Dataset(tissue,site)","Var1","Var2","rho(spearman)","p-value")
colnames(pcs_vs_qc_var_report) = c("Dataset(tissue,site)","PC",
  "qc_metric","rho(spearman)","p-value")
pcs_vs_qc_var_report[,5] = format(
  as.numeric(pcs_vs_qc_var_report[,5]),digits=3)
pcs_vs_qc_var_report[,4] = format(
  as.numeric(pcs_vs_qc_var_report[,4]),digits=3)
metadata_variable_assoc_report[,5] = format(
  as.numeric(metadata_variable_assoc_report[,5]),digits=3)
metadata_variable_assoc_report[,4] = format(
  as.numeric(metadata_variable_assoc_report[,4]),digits=3)
```

## Print the significant results in tables
```{r}
kable(pcs_vs_qc_var_report,longtable=T,
      caption = "Correlations between PCs and other variables")  %>%
  kable_styling(latex_options = c("repeat_header"),font_size = 8)
```

# Print all flagged samples

```{r}
# add the dataset name to mop outliers
sample2dataset = paste(tolower(rnaseq_meta[,c("Tissue")]),
        tolower(rnaseq_meta[,c("GET_site")]),sep=",")
names(sample2dataset) = rownames(rnaseq_meta)
#colnames(sex_check_outliers) = c("dataset","sample")

all_flagged = unique(c(
  #mop_flagged_samples,
  #sex_check_outliers[,"sample"],
  pca_outliers_report[,"sample"]
))

flagged_sample_report = c()
for(samp in all_flagged){
  samp_dataset = sample2dataset[samp]
  samp_metric = NULL
  # if(samp %in% mop_flagged_samples){
  #   samp_metric = "MOP-flagged"
  # }
  # if(samp %in% sex_check_outliers[,"sample"]){
  #   samp_metric = c(samp_metric,"Sex-flagged")
  # }
  if(samp %in% pca_outliers_report[,"sample"]){
    curr_pcs = pca_outliers_report[pca_outliers_report[,"sample"]==samp,"PC"]
    samp_metric = c(samp_metric,curr_pcs)
  }
  flagged_sample_report = rbind(flagged_sample_report,
      c(samp,samp_dataset,paste(samp_metric,collapse=","))
  )
}
colnames(flagged_sample_report) = c("Vial label","Dataset","Methods")
kable(flagged_sample_report,longtable=T,
      caption = "Outliers detected and by tissue PCA data")%>%
  kable_styling(font_size = 8,latex_options = c("hold_position", "repeat_header"))

```


## Save the results in the bucket

```{r}
rnaseq_bic_qc_reports = list(
  #dmaqc_data_sex_vs_times = dmaqc_data_sex_vs_times,
  metadata_variable_assoc_report = metadata_variable_assoc_report,
  #mop_flagged_samples,
  pca_outliers_report,
  pcs_vs_qc_var_report=pcs_vs_qc_var_report
)
save_to_bucket(rnaseq_bic_qc_reports,file = "rnaseq_bic_qc_reports.RData",
               bucket = "gs://bic_data_analysis/pass1a/rnaseq/")
save(rnaseq_bic_qc_reports, file="/oak/stanford/groups/smontgom/nicolerg/MOTRPAC/PASS_ANALYSIS/PASS1B/RNA/differential_splicing/bic_qc_report/rdata/rnaseq_bic_qc_reports.RData")
```

